Adds docs for `dictGet` commands.

johnnymatthews opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- Documentation (changelog entry is not required)

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

---

Adds documentation for `dictGetXYZ` commands. Also add a template that folks can use to create new functions docs.

Closes https://github.com/ClickHouse/clickhouse-docs/issues/1886.

johnnymatthews committed:
Just adds headers for dictGet commands.

johnnymatthews committed:
Adds int and uint versions of dictGet function.

johnnymatthews committed:
Merge branch 'master' into docs/adds-dictGetXYZ

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #59392 from ClickHouse/ClickHouse:docs/adds-dictGetXYZ -->
*This is an automated comment for commit 0c1bec1989413564032255aeba427cb17795839b with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59392/0c1bec1989413564032255aeba427cb17795839b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

johnnymatthews committed:
Adds Floats, Dates, UUID, String, and IPs to getDict.

johnnymatthews committed:
Adds template for adding function docs.

This template isn't in any kind of automation or CI/CD process. It's just here to make it easier for docs writers to use the same layout when creating function documentation.

johnnymatthews committed:
Adds dictGet, dictGetOrDefault, and dictGetOrNull.

johnnymatthews committed:
Cleans up function template.

johnnymatthews committed:
Adds remaining `...OrDefault` functions.

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+**Implementation details**
+
+This function will throw an exception if the key does not exist in the dictionary.
```
This is not true, this function will throw an exception if it cannot parse the value of the attribute or the value does not match the attribute data type, not if the key does not exist.

Avogar replied:
It will also throw an exception when attribute is not `Int8`

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -7,226 +7,188 @@ sidebar_label: Dictionaries
+## dictGetOrDefault
 
-Create a text file `ext-dict-mult.csv` containing the following:
+Retrieves values from a dictionary. If that value cannot be found, returns the default value set by the user.
```
I think it's not good to say `If that value cannot be found`, because we search values by keys, and if we cannot find the key, then the default value will be set

johnnymatthews replied:
I don't understand. Please leave a suggestion.

Avogar replied:
```suggestion
Retrieves values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -7,226 +7,188 @@ sidebar_label: Dictionaries
-Input table:
+If there is no the key, corresponding to `id_expr`, in the dictionary, then returns the value passed as the `default_value_expr` parameter.
+
+**Implementation details**
+
+None.
+
```
If none, please, remove this section

johnnymatthews replied:
I disagree. Pointing out that there aren't any specific implementation details is useful. Otherwise it could look like we just forgot to put the details in (which has happened all throughout our docs).

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -7,226 +7,188 @@ sidebar_label: Dictionaries
+LIFETIME(MIN  0 MAX  1000)
+LAYOUT(FLAT());
+
+SELECT dictGetOrNull('id_value_dictionary', 'missing_attribute', toUInt64(1)) AS value;
```
It will throw an exception `Code: 36. DB::Exception: No such attribute 'missing_attribute': While processing dictGetOrNull('id_value_dictionary', 'missing_attribute', toUInt64(1)) AS value. (BAD_ARGUMENTS)`. Please, test that your examples works.

Functions `OrNull/OrDefault` returns default value/null when the requested key doesn't exist, not when

johnnymatthews replied:
Suggestions are welcome.

![image](https://github.com/ClickHouse/ClickHouse/assets/9611008/5aa633ab-07b3-4061-ba6e-651daafc3381)

Avogar replied:
```suggestion
SELECT dictGetOrNull('id_value_dictionary', 'value', toUInt64(3)) AS value;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
 
-        - `dictGet[Type]` returns the content of the `<null_value>` element specified for the attribute in the dictionary configuration.
-        - `dictGet[Type]OrDefault` returns the value passed as the `default_value_expr` parameter.
+This function will throw an exception if the key does not exist in the dictionary.
```
This is not true, it's `OrDefault` function, it will return default value in this case. But it may throw an excetpion if attribute is not `Int8` type or if default value is not `Int8` or cannot be parsed as `Int8`

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetInt8OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;
```
This query will throw an exception that it cannot parse `'Default return string.'` as `Int8`

Avogar replied:
```suggestion
SELECT dictGetInt8OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

And the same for all other `dictGet<type>OrDefault`

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
````diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+'Default return string.'
+```
+
+## dictGetInt16
````
Do we really need this copy paste for every function `dictGet<someType>`? I don't think it's necessary, the are all the same and we can make some general documentation for `dictGet<someType>` and `dictGet<someType>OrDefault`, we can just list supported types and add several examples

johnnymatthews replied:
Search engine optimization.

Avogar replied:
What optimizations? I don't see why we need huge copy paste in docs with the same text and examples

johnnymatthews replied:
If a user searches for `dictGetInt16`, that means they're looking for documentation for the function `dictGetInt16`. Not Int8, not Float32, not anything else. They know what they want, and what they want is `dictGtiInt16`.

In the future, we can automate some of this documentation to populate the changes for us. But right now, the docs team's goal is to document all the holes left by engineers of old. That's what we're doing.

Avogar replied:
I still think if the user searches for `dictGetInt16` and finds general documentation about `dictGet<type_name>`, it will be enough to understand everything about `dictGetInt16`. For me having 2k lines of documentation for each `dictGet<type_name>` is not a good documentation. But I can be wrong, let me ask what my colleagues think about it.

johnnymatthews committed:
Update docs/en/sql-reference/functions/ext-dict-functions.md

Co-authored-by: Kruglov Pavel <48961922+Avogar@users.noreply.github.com>

johnnymatthews committed:
Update docs/en/sql-reference/functions/ext-dict-functions.md

Co-authored-by: Kruglov Pavel <48961922+Avogar@users.noreply.github.com>

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -7,226 +7,188 @@ sidebar_label: Dictionaries
-    dictGetOrNull('range_key_dictionary', 'value_nullable', number, toDate('2019-05-20')),
-    dictGetOrNull('range_key_dictionary', ('value', 'value_nullable'), number, toDate('2019-05-20'))
-FROM system.numbers LIMIT 5 FORMAT TabSeparated;
+Retrieves values from a dictionary. If that value cannot be found, returns null. This is different to `dictGet` which throws an exception if the value cannot be found.
```
```suggestion
Retrieves values from a dictionary. Returns `NULL` for keys that are not found in the dictionary. This is different to `dictGet` which throws an exception if the key cannot be found.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetInt8OrDefault
+
+Retrieves type Int8 values from a dictionary. If the Int8 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type Int8 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetInt16OrDefault
+
+Retrieves type Int16 values from a dictionary. If the Int16 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type Int16 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetInt16OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;
```
```suggestion
SELECT dictGetInt16OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetInt32OrDefault
+
+Retrieves type Int32 values from a dictionary. If the Int32 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type Int32 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetInt32OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;
```
```suggestion
SELECT dictGetInt32OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetInt64OrDefault
+
+Retrieves type Int64 values from a dictionary. If the Int64 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type Int64 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;
```
```suggestion
SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetUInt8OrDefault
+
+Retrieves type UInt8 values from a dictionary. If the UInt8 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type UInt8 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetUInt8OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;
```
```suggestion
SELECT dictGetUInt8OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetUInt16OrDefault('country_codes', 'code_designation', '999', 'Default value string.') AS code_designation_output;
```
```suggestion
SELECT dictGetUInt16OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetUInt32OrDefault
+
+Retrieves type UInt32 values from a dictionary. If the UInt32 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type UInt32 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetUInt32OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;
```
```suggestion
SELECT dictGetUInt32OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetUInt64OrDefault
+
+Retrieves type UInt64 values from a dictionary. If the UInt64 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type UInt64 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '2', 'Default return string.') AS code_designation_output;
```
```suggestion
SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetFloat32OrDefault
+
+Retrieves type Float32 values from a dictionary. If the Float32 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type Float32 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetFloat32OrDefault('country_codes', 'float_designation', '999', 'Default return string.') AS dict_output;
```
```suggestion
SELECT dictGetFloat32OrDefault('country_codes', 'float_designation', '999', 42.42) AS dict_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetFloat64OrDefault
+
+Retrieves type Float64 values from a dictionary. If the Float64 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type Float64 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetFloat64OrDefault('country_codes', 'float_designation', '999', 'Default return string.') AS dict_output;
```
```suggestion
SELECT dictGetFloat64OrDefault('country_codes', 'float_designation', '999', 42.42) AS dict_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetDateOrDefault
+
+Retrieves type Date values from a dictionary. If the Date attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type Date values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetDate('country_codes', 'independence_day', '100', 'Default return string.') AS dict_output;
```
```suggestion
SELECT dictGetDateOrDefault('country_codes', 'independence_day', '100', '1970-01-01') AS dict_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetDateTimeOrDefault
+
+Retrieves type DateTime values from a dictionary. If the DateTime attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type DateTime values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetUUIDOrDefault
+
+Retrieves type UUID values from a dictionary. If the UUID attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type UUID values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetUUIDOrDefault('countries_dict', 'special_id', 999, 'Default return string.') AS dict_output;
```
```suggestion
SELECT dictGetUUIDOrDefault('countries_dict', 'special_id', 999, '00000000-0000-0000-0000-000000000000') AS dict_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetStringOrDefault
+
+Retrieves type String values from a dictionary. If the String attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type String values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetIPv4OrDefault
+
+Retrieves type IPv4 values from a dictionary. If the IPv4 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type IPv4 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetIPv4OrDefault('countries_dict', 'ip_address', 999, 'Default return string.') AS dict_output;
```
```suggestion
SELECT dictGetIPv4OrDefault('countries_dict', 'ip_address', 999, '0.0.0.0') AS dict_output;
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+
+## dictGetIPv6OrDefault
+
+Retrieves type IPv6 values from a dictionary. If the IPv6 attribute cannot be found, returns the specified default value instead.
```
```suggestion
Retrieves type IPv6 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

Avogar commented on docs/en/sql-reference/functions/ext-dict-functions.md:
```diff
@@ -481,44 +443,1954 @@ SELECT dictGetAll('regexp_dict', 'tag', 'foobarbaz', 2);
+LAYOUT(FLAT())
+LIFETIME(MIN 0 MAX 1000);
+
+SELECT dictGetIPv6OrDefault('countries_dict', 'ip_address', 999, 'Default return string.') AS dict_output;
```
```suggestion
SELECT dictGetIPv6OrDefault('countries_dict', 'ip_address', 999, '::ffff:0.0.0.0') AS dict_output;
```

Avogar commented:
@johnnymatthews hi! Will you continue this PR? There are a lot of comments, but they are almost all the same with small changes
